ktlint-ruleset-standard/src/main/kotlin/com/pinterest/ktlint/ruleset/standard/TrailingCommaOnDeclarationSiteRule.kt

```diff
@@ -0,0 +1,191 @@
+            isTrailingCommaAllowed = node.isTrailingCommaAllowed(),
+            autoCorrect = autoCorrect,
+            emit = emit
+        )
+    }
+
+    private fun visitValueList(
+        node: ASTNode,
+        autoCorrect: Boolean,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit
+    ) {
+        if (node.treeParent.elementType != ElementType.FUNCTION_LITERAL) {
+            node
+                .children()
+                .lastOrNull { it.elementType == ElementType.RPAR }
+                ?.let { inspectNode ->
+                    node.reportAndCorrectTrailingCommaNodeBefore(
+                        inspectNode = inspectNode,
+                        isTrailingCommaAllowed = node.isTrailingCommaAllowed(),
+                        autoCorrect = autoCorrect,
+                        emit = emit
+                    )
+                }
+        }
+    }
+
+    private fun visitTypeList(
+        node: ASTNode,
+        autoCorrect: Boolean,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit
+    ) {
+        val inspectNode = node
+            .children()
+            .first { it.elementType == ElementType.GT }
+        node.reportAndCorrectTrailingCommaNodeBefore(
+            inspectNode = inspectNode,
+            isTrailingCommaAllowed = node.isTrailingCommaAllowed(),
+            autoCorrect = autoCorrect,
+            emit = emit
+        )
+    }
+
+    private fun visitWhenEntry(
+        node: ASTNode,
+        autoCorrect: Boolean,
+        emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit
+    ) {
+        val psi = node.psi
+        require(psi is KtWhenEntry)
+        if (psi.isElse || psi.parent.cast<KtWhenExpression>().leftParenthesis == null) {
+            // no commas for "else" or when there are no opening parenthesis for the when-expression
+            return
+        }
+
+        val inspectNode = node
+            .children()
+            .first { it.elementType == ElementType.ARROW }
+        node.reportAndCorrectTrailingCommaNodeBefore(
+            inspectNode = inspectNode,
+            isTrailingCommaAllowed = node.isTrailingCommaAllowed(),
+            autoCorrect = autoCorrect,
+            emit = emit
+        )
+    }
+
+    public companion object {
+
+        internal const val ALLOW_TRAILING_COMMA_NAME = "ij_kotlin_allow_trailing_comma"
+        private const val ALLOW_TRAILING_COMMA_DESCRIPTION = "Defines whether a trailing comma (or no trailing comma)" +
+            "should be enforced on the defining side," +
+            "e.g. parameter-list, type-argument-list, lambda-value-parameters, enum-entries, etc."
+
+        private val BOOLEAN_VALUES_SET = setOf("true", "false")
+
+        // TODO: Rename property to trailingCommaOnDeclarationSite. The word 'allow' is misleading as the comma is
+        //       enforced when the property is enabled and prohibited when disabled.
```

Comment:
The TODO comments are nice to be fixed - But I'd prefer to not to extend the scope of this PR